### PR TITLE
feat: shared elasticsearch

### PR DESCRIPTION
**Pull request by keithgg**
## Description

This PR adds the helm chart to deploy a standalone Elasticsearch cluster. Ref overhangio/tutor-forum#4 

When `elasticsearch.enabled` is true it will deploy a cluster with 3 replicas by default.

Operators will need to add/update the following setting in each of their instances. 

```yaml
K8S_HARMONY_ENABLE_SHARED_ELASTICSEARCH: true
RUN_ELASTICSEARCH: false
```

## Caveats

- In order for SSL to work without warnings the CA certificate needs to be mounted in the relevant pods. This is out of scope for this PR as it needs to be an instance level change, which requires more work than intended.
- It's not possible to mount the CA certificate for the forum due to https://github.com/overhangio/tutor/issues/791

## Testing instructions

- Follow the instructions to install the chart and rebuild the Open edX image (`tutor image build openedx`). 
  - Or use my prebuilt docker image: `DOCKER_IMAGE_OPENEDX: docker.io/keithgg/openedx:15.1.2`
- Add the configuration setting: `tutor config save --set K8S_HARMONY_ENABLE_SHARED_ELASTICSEARCH=true --set RUN_ELASTICSEARCH=false`
- Run the command `tutor harmony create-elasticsearch-user` to set up your user on the ES cluster
- `tutor k8s start && tutor k8s init`

Anything ES-related should work as normal, specifically importing a course and doing a search. You can then check that the indexes are created by running the following on one of the ES pods:

```shell
$ curl --insecure -u elastic:${ELASTIC_PASSWORD} https://localhost:9200/_cat/indices
green open .geoip_databases       22H5xGxATfexbL9sos_jQA 1 1 41 0 78.7mb 39.3mb
green open openedx-02-course_info m4cKRBuqReOhKMt2OLWzTA 1 1  0 0   452b   226
```

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @keithgg!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by e0d** [comment]
@bradenmacdonald are you the appropriate reviewer?

**Comment by bradenmacdonald** [comment]
@e0d yes I am planning to review this. But for this particular project we had said we'd like to have each PR reviewed by at least one other provider, so before it merges I'd like to have a review from someone outside of OpenCraft.

@MoisesGSalas would you be interested in reviewing this?

**Comment by MoisesGSalas** [comment]
@bradenmacdonald sure, I will take a look as soon as possible.

**Comment by bradenmacdonald** [comment]
> Assuming you instance name is openedx-01, we'll need to generate a bcrypt hash for the user's ES password.

> Add to your values.yml
> ```
> ...
>  users:
>    openedx-01: "{the bcrypt hash}"
> ```

@keithgg Maybe I'm missing something, but we need to avoid including any instance-specific setup in the chart and its `values.yaml`. So we can configure a "root" ElasticSearch user/password there, but if each Open edX instance needs to have its own credentials and to create its own ES user, that should happen in Tutor (probably in this repo's included Tutor plugin). Because we want the use of this chart to set up the cluster to be mostly a one-time thing, and then just run a Tutor or Grove command for each additional instance that gets installed.

**Comment by keithgg** [comment]
@bradenmacdonald Ah good shout. In my head I was thinking in terms of Grove we you had an extra layer of software that will do the config for you.

Now that I think about it, it might be possible to add the users via the ES API as part of an init task. I'll check and revert back.

**Comment by itsjeyd** [comment]
Hey @keithgg, any updates on when you're planning to get back to this PR?

**Comment by keithgg** [comment]
@bradenmacdonald I've made the changes to simplify the setup. I created a command to set up the ElasticSearch auth that the operator will need to run once.

There wasn't a tutor hook that I could use to run it automatically, since all the current hooks are set up to run within a specific container or against a specific namespace.

**Comment by bradenmacdonald** [comment]
Thanks @keithgg, I'll try to review in the next couple of days. Heads up that this will need to be rebased after #10 merges; I tried to merge it just now but don't have rights. Shouldn't affect most of the main focus of this PR though.

File: tutor-multi-chart/Chart.yaml
Comment (bradenmacdonald) [inline tutor-multi-chart/Chart.yaml:24]: As discussed in the call yesterday - we may need to change this to OpenSearch or something else in the future.

**Comment by bradenmacdonald** [comment]
> There wasn't a tutor hook that I could use to run it automatically, since all the current hooks are set up to run within a specific container or against a specific namespace.

How does this work for using shared MySQL, do you have to create the instance-specific credentials manually in that case?

File: tutor-multi-chart/values.yaml
Comment (bradenmacdonald) [inline tutor-multi-chart/values.yaml:84]: The comment says "The chart will handle certificate creation" but `createCert: false` sounds like we're telling the chart _not_ to handle it?

File: README.md
Comment (bradenmacdonald) [inline README.md:155]: Are the two caveats mentioned in the PR description still relevant?

> * In order for SSL to work without warnings the CA certificate needs to be mounted in the relevant pods. This is out of scope for this PR as it needs to be an instance level change, which requires more work than intended.
> * It's not possible to mount the CA certificate for the forum due to https://github.com/overhangio/tutor/issues/791

If so, please add them to the README here.

File: tutor-contrib-multi-plugin/tutor_multi_k8s_plugin/elasticsearch.py
Comment (bradenmacdonald) [inline tutor-contrib-multi-plugin/tutor_multi_k8s_plugin/elasticsearch.py:4]: Nit: please add a docstring to this class.

**Review by bradenmacdonald — COMMENTED** [review]
@keithgg awesome work here 💯 

Thanks for figuring this out and for working through the related issues in tutor and edx-search.

I'm happy with the code, just had some minor requests/comments. There are now quite a few merge conflicts so if you could please rebase it, I'll do a quick test on the rebased version then it's good to go from my perspective.

Then I'll approve it and hand it over to @MoisesGSalas  for a final review.

**Comment by keithgg** [comment]
Thanks for the review @bradenmacdonald. I'll go through you comments and make the necessary changes tomorrow.

File: tutor-multi-chart/Chart.yaml
Comment (keithgg) [reply tutor-multi-chart/Chart.yaml:29]: This configuration for `OpenSearch` is similar, but not the same. At minimum the names of the config values are OpenSearch and not ElasticSearch. Configuring users might also be different, but I didn't check in detail.

I'll create a dedicated ticket for it.

File: tutor-multi-chart/values.yaml
Comment (keithgg) [reply tutor-multi-chart/values.yaml:34]: Ah thanks for pointing that out. The wording was confusing. I meant that the `k8s-harmony` chart would create the certs.

**Comment by keithgg** [comment]
> > There wasn't a tutor hook that I could use to run it automatically, since all the current hooks are set up to run within a specific container or against a specific namespace.
> 
> How does this work for using shared MySQL, do you have to create the instance-specific credentials manually in that case?

@bradenmacdonald there isn't a shared MySQL implementation? Did you mean something else?

**Comment by bradenmacdonald** [comment]
> I'm not sure what you mean here? There isn't a shared MySQL implementation (unless I'm mistaken).

What I meant was that [you can use Tutor with an existing shared MySQL cluster as described in the docs](https://docs.tutor.overhang.io/configuration.html#mysql), and it requires a root user/password (`MYSQL_ROOT_USERNAME `/`MYSQL_ROOT_PASSWORD`) which can be per-instance or shared by all instances, but in any case [init code like this](https://github.com/overhangio/tutor/blob/c26999ec658a216b68797e2d38c522ff53e7fbd3/tutor/templates/jobs/init/mysql.sh#L19) will then create service-specific users.

Can we not do the same thing? It doesn't really matter which container or namespace the `init` script runs in, as long as it can connect to the shared ElasticSearch DB and provision a new user account over the network. But if that approach is not secure enough, we can do something else.

File: README.md
Comment (bradenmacdonald) [inline README.md:167]: This is now merged, right?

File: README.md
Comment (bradenmacdonald) [inline README.md:166]: Might be confusing to say "out of scope of this PR" in the README since it won't be part of a PR once merged... maybe say "this isn't yet implemented" ?

**Review by bradenmacdonald — APPROVED** [review]
Thanks @keithgg , I think this is a great start.

**Comment by bradenmacdonald** [comment]
@MoisesGSalas Once you are back and have time, could you please do the final review on this? We can merge it once you've approved it too.

File: tutor-contrib-multi-plugin/tutor_multi_k8s_plugin/plugin.py
Comment (MoisesGSalas) [inline tutor-contrib-multi-plugin/tutor_multi_k8s_plugin/plugin.py]: While I was testing the string generated included uppercase characters so when I tried re-indexing a course I got an 'invalid index name must be lowercase' error. Maybe we should normalize this string?

**Review by MoisesGSalas — COMMENTED** [review]
Sorry for the late response @keithgg, I left a few comments so you can take a look when you have a chance.

File: README.md
Comment (keithgg) [reply README.md:167]: Yep. I don't want to muddy the waters to much, so are you happy to keep this PR as is and I'll create an issue to update?

File: tutor-contrib-multi-plugin/tutor_multi_k8s_plugin/plugin.py
Comment (keithgg) [reply tutor-contrib-multi-plugin/tutor_multi_k8s_plugin/plugin.py]: Ah, I got lucky that it generated a lowercase string for me. I've updated the filter.

File: README.md
Comment (bradenmacdonald) [reply README.md:167]: Works for me :+1:

**Review by MoisesGSalas — APPROVED** [review]
Seems like it's working fine now. Thanks a lot @keithgg.

File: README.md
Comment (keithgg) [reply README.md]: Sure @bradenmacdonald, make sense. I've clarified the wording.

**Comment by keithgg** [comment]
Thank you for the review @MoisesGSalas!

> > I'm not sure what you mean here? There isn't a shared MySQL implementation (unless I'm mistaken).
> 
> What I meant was that [you can use Tutor with an existing shared MySQL cluster as described in the docs](https://docs.tutor.overhang.io/configuration.html#mysql), and it requires a root user/password (`MYSQL_ROOT_USERNAME `/`MYSQL_ROOT_PASSWORD`) which can be per-instance or shared by all instances, but in any case [init code like this](https://github.com/overhangio/tutor/blob/c26999ec658a216b68797e2d38c522ff53e7fbd3/tutor/templates/jobs/init/mysql.sh#L19) will then create service-specific users.
> 
> Can we not do the same thing? It doesn't really matter which container or namespace the `init` script runs in, as long as it can connect to the shared ElasticSearch DB and provision a new user account over the network. But if that approach is not secure enough, we can do something else.

@bradenmacdonald we can, but that means we'll need to add explicit steps for the user to retrieve the admin password and add it to their `config.yml`. My earlier understanding was that we wanted to limit the number of setup steps users need to run so I automated as much as I could.

If you're happy, I'll add this as part #23, since we'll need to retrieve the CA cert from the `tutor-multi` as well.

**Comment by itsjeyd** [comment]
@keithgg Is this PR blocked on #23 now, or could we merge it once tests are green (since @bradenmacdonald and @MoisesGSalas already approved the changes)?

**Comment by keithgg** [comment]
@itsjeyd this can be merged. The remaining work will take place in #23. I don't have permissions to merge though. @bradenmacdonald are you able to?

**Comment by bradenmacdonald** [comment]
Done!

**Pull request merged**
